### PR TITLE
fix: try to fix issue #2585 by adding block cleanup from db directly.

### PR DESCRIPTION
**Pull request by daogangtang**
fix: try to fix issue #2585 by adding block cleanup from db directly.

### CODE REVIEW

File: chain/src/chain.rs [hunk 1]
```diff
@@ -1065,6 +1065,49 @@ impl Chain {
 			count, tail.height
 		);
 
+		// Remove short live fork blocks
+		// Get all block headers in db
+		let all_block_headers = self.store.all_block_headers()?;
```
Comment (cadmuspeverell) [inline chain/src/chain.rs]: You'll want to iterate over all blocks, not block headers. We never compact/delete the block headers, so using your approach would end up iterating over 100s of thousands of headers. Blocks are compacted frequently, which means we never have more than a few weeks worth of them. Just iterate over each block in the store, check `is_on_current_chain` and check block.header.height >= tail.height. If either of those is false, call delete_block.

Comment (daogangtang) [reply chain/src/chain.rs]: Hi, updated with new algorithm u advised :)

File: chain/src/store.rs [hunk 2]
```diff
@@ -378,6 +378,13 @@ impl<'a> Batch<'a> {
 			db: self.db.child()?,
 		})
 	}
+
+	/// An iterator to all lived block now
+	pub fn iter_lived_blocks(&self) -> Result<SerIterator<Block>, Error> {
```
Comment (antiochp) [inline chain/src/store.rs]: Maybe rename this to `blocks_iter()` for consistency with `difficulty_iter()` (and others)?

Comment (daogangtang) [reply chain/src/store.rs]: Ok.

File: chain/src/chain.rs [hunk 3]
```diff
@@ -1026,38 +1026,18 @@ impl Chain {
...
-			}
-			if current.height <= 1 {
-				break;
-			}
-			match batch.get_previous_header(&current) {
-				Ok(h) => current = h,
-				Err(NotFoundErr(_)) => break,
-				Err(e) => return Err(From::from(e)),
+		// Remove old blocks and short live fork blocks
+		// here b is a lived block
```
Comment (antiochp) [inline chain/src/chain.rs]: I'm not sure what you mean by "lived" block - maybe reword these comments?

Comment (daogangtang) [reply chain/src/chain.rs]: Ok.

File: chain/src/chain.rs [hunk 4]
```diff
@@ -1026,38 +1026,18 @@ impl Chain {
...
+		let _ = batch.iter_lived_blocks()?.map(|(_, b)| {
```
Comment (antiochp) [inline chain/src/chain.rs]: This can be simplified to - 

```
for (_, b) in batch.iter_lived_blocks()? {
    ...
}
```

Comment (daogangtang) [reply chain/src/chain.rs]: Ok.

**Comment by antiochp** [comment]
Looks like lmdb is locking up internally with this change - our "batch" is very susceptible to deadlocking with multiple accesses.
I'm looking to see if there are any obvious culprits here.

**Comment by daogangtang** [comment]
> Looks like lmdb is locking up internally with this change - our "batch" is very susceptible to deadlocking with multiple accesses.
> I'm looking to see if there are any obvious culprits here.

Wow, this is my first attempt to contribute to grin, so lack experiences on it. :(

**Comment by daogangtang** [comment]
Or should I use store_instance.iter() rather than batch.iter() in exported iterator in chain/src/store.rs?

**Comment by antiochp** [comment]
Ha - It's the call to `is_on_current_chain()`. 
Turns out we cannot actually call this inside the iterator (as it tries to get a lock on the db but the batch already has a lock on it).
I'll take another pass later today and see if we can rework this to use the existing batch.

**Comment by daogangtang** [comment]
In my opinion, maybe we don't need batch lock in the procedure of iterating over blocks in db, this is not necessary for the whole block cleanup, only necessary when delete one block. So, we can add an iterator to ChainStore implementation, rather than Batch.

**Comment by antiochp** [comment]
I think we actually got the logic wrong here.

Now that we are iterating over _all_ blocks in the db, all we really care about is cleaning up blocks older than the threshold.

So we do not need to actually check for `is_on_current_chain()` - if its older than `tail.height` we simply delete it, whether on current chain or not.

And this should resolve the deadlocking issue as well I think.

**Comment by daogangtang** [comment]
The height of a fork block MUST be smaller than tail.height? Surely?

**Comment by antiochp** [comment]
> The height of a fork block MUST be smaller than tail.height? Surely?

Yes - we want to delete block older (i.e. lower height) than tail.height
